File: pulsar/java/src/main/java/com/examples/hello/pulsar/services/PulsarRunner.java
package com.examples.hello.pulsar.services;

import lombok.Getter;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public abstract class PulsarRunner implements AutoCloseable, PulsarRunnable {
  @Getter
  private boolean running;

  @Override
  public void stop() {
    running = false;
  }

  @SneakyThrows
  public void run() {
    if (this.isRunning()) throw new UnsupportedOperationException("Already started");
    running = true;
    while (running) {
      try {
        callOnce();
      } catch (Exception e) {
        log.warn("Could not process", e);
      }
    }
  }

  protected abstract void callOnce() throws Exception;
}

File: pulsar/kotlin/src/main/kotlin/com/example/hello/pulsar/services/PulsarRunner.kt
package com.example.hello.pulsar.services

import org.slf4j.LoggerFactory

abstract class PulsarRunner : AutoCloseable, PulsarRunnable {
    override var isRunning = false

    override fun stop() {
      isRunning = false
    }

    override fun run() {
        if (isRunning) throw UnsupportedOperationException("Already started")
        isRunning = true
        while (isRunning) {
            try {
                callOnce()
            } catch (e: Exception) {
                log.warn("Could not process", e)
            }
        }
    }

    protected abstract fun callOnce()

    companion object {
        private val log = LoggerFactory.getLogger(PulsarRunner::class.java)
    }
}
